rfcs/0012-pipeline-executor.md

```diff
@@ -0,0 +1,367 @@
+    }
+    mod_config[mods[2]] = mconfig3
+``` 
+
+### 3.3. Use pipeline_executor to build pipeline module with the said subgraph and configuration.
+
+Pipeline executor provide a build function to compile and save the compile output into disk,
+following is a example
+
+```python
+    with relay.build_config(opt_level=3):
+        pipeline_mods, string_config = pipeline_executor.build_pipeline(
+            mod_config, "<path to storage the build output>"
+        )
+
+```
+
+### 3.4. Use pipeline_executor to load pipeline module to run network in pipeline parallism mode.
+
+Pipeline executor working at asynchronous mode, unlike the synchronous mode set_input->run->get_out
+sequence, Pipeline executor execte in such order for example.
+    set_input--> run
+    set_input--> run
+    get_ouput
+    set_input-->run
+    get_output
+    get_output
+
+get_output can be called before set_input/run, and it may return a empy array that means no output ready
+or a valid output.
+
+following is one example
+
+```python
+#...
+
+def get_output(outputs, module):
+  suc = False
+  output = pipeline_module.get_output()
+  if len(output):
+    curOutputs = [output.asnumpy() for data in output]
+    outputs.append(curOutputs)
+    suc = True
+
+  return suc
+    
+
+pipeline_outputs = []
+datas = []
+for i in range(len(mods) + 1):
+  datas.append(np.full(dshape, 3 + i).astype("float32"))
+pipeline_module = pipeline_executor.create(pipeline_mods, string_config)
+
+for data in datas:
+    get_output(pipeline_outputs, pipeline_module)
+    pipeline_module.set_input("data_0", data)
+    pipeline_module.set_input("data_1", data, mod_idx=2)
+    pipeline_module.run()
+    get_output(pipeline_outputs, pipeline_module)
+
+left = len(datas) - len(pipeline_outputs)
+while(left > 0):
+    left = left - 1 if get_output(pipeline_outputs, pipeline_module) else left
+```
+
+## 4 Reference-level explanation
+This section introduces the underlying techniques for the pipeline executor.
+The figure below briefly illustrates the workflow of the system
+
+Pipeline executor architecture
+![meta-schedule-workflow](../resources/pipeline-executor-arch.png)
+
+Manually construct the subgraph
+![meta-schedule-workflow](../resources/pipeline-executor-subgraph-split.png)
+
+How pipeline executor runtime work
+![meta-schedule-workflow](../resources/pipeline-executor-runtime.png)
+
+The pipeline executor schedule logic
+![meta-schedule-workflow](../resources/pipeline-executor-schedule.png)
+
+The network pipeline compute effect
+![meta-schedule-workflow](../resources/pipeline-executor-pipeline.png)
+
+
+## 5. Drawbacks
+
+
+Pipeline executor currently need manually subgraph split and configuration construct,
+Further graph split feature would doing automatically split.
+
+## 6. Rationale and alternative
+
+
+Compute graph can get split into subgraph and pipeline execution can implement parallism
+when these subgraph have dependency relation.
```

Comment:
This statement is not related to "Rationale and alternative". In this section, you should include other possible solutions without pipeline executor to illustrate that pipeline executor is really necessary to achieve the goal.